File: Etl/etl/extract.py
import pandas as pd
import mysql.connector
from dotenv import load_dotenv
import os
import glob

load_dotenv()

def recuperer_donnees(id_nuit : int):
    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    chemin_csv = os.path.join(base_dir, "raw", "traite", f"signal-psg-patient-{id_nuit}-nuit-{id_nuit}.csv")

    df_capteur = pd.read_csv(chemin_csv)

    # lecture des événements (MySQL)
    # Change the connection block in extract.py to this:
    conn = mysql.connector.connect(
        host=os.environ.get("DB_HOST"),
        port=os.environ.get("DB_PORT"),
        user=os.environ.get("DB_USER"),
        password=os.environ.get("DB_PASSWORD"),
        database=os.environ.get("DB_NAME")
    )

    cur = conn.cursor(buffered=True)
    cur.execute('SELECT * FROM evenement_respiratoire WHERE id_nuit = %s', (id_nuit,))
    df_events = cur.fetchall()

    # Des bugs apparaissent lorsque chaque query est execute séparement
    # recours a une fnction local pour fix
    def count_events(cursor, id_nuit, types=None):
        if types:
            placeholders = ", ".join(["%s"] * len(types))
            cursor.execute(
                f"SELECT COUNT(*) FROM evenement_respiratoire WHERE id_nuit = %s AND type_evenement IN ({placeholders})",
                (id_nuit, *types)
            )
        else:
            cursor.execute(
                "SELECT COUNT(*) FROM evenement_respiratoire WHERE id_nuit = %s",
                (id_nuit,)
            )
        return cursor.fetchone()

    nbapnees = count_events(cur, id_nuit, ["apnée obstructive", "apnée centrale"])
    nbhypopnees = count_events(cur, id_nuit, ["hypopnée"])
    nbrera = count_events(cur, id_nuit, ["RERA"])
    nbr_events = count_events(cur, id_nuit)

    queries_response = {
        'df_capteur': df_capteur,
        'df_events': df_events,
        'nbapnees': nbapnees,
        'nbhypopnees': nbhypopnees,
        'nbrera': nbrera,
        'nbr_events': nbr_events,
    }
    print(f"  apnées={nbapnees[0]}, hypopnées={nbhypopnees[0]}, RERA={nbrera[0]}, total={nbr_events[0]}")
    conn.close()

    return queries_response

File: Api/etl/extract.js
import fs from 'fs';
import path from 'path';
import mysql from 'mysql2/promise';
import csvParser from 'csv-parser';



const pool = mysql.createPool({
    host: process.env.DB_HOST || 'localhost',
    user: process.env.DB_USER || 'root',
    port: process.env.DB_PORT || 3333,
    password: process.env.DB_PASSWORD || 'root',
    database: process.env.DB_NAME || "clinique2nuitsv2",
    waitForConnections: true,
    connectionLimit: 10
});

export async function recupererDonnees(idNuit) {

    const baseDir = 'C:/python-projs/Clinique_d_Arles/server';

    const cheminCsv = path.join(baseDir, "raw", "traite", `signal-psg-patient-${idNuit}-nuit-${idNuit}.csv`);


const dfCapteur = await new Promise((resolve, reject) => {

    if (!fs.existsSync(cheminCsv)) {
        return reject(new Error(`Fichier introuvable : ${cheminCsv}`));
    }

    const results = [];
    fs.createReadStream(cheminCsv)//permet de lire un fichier par morceaux (flux) fs (file system)
        .on('error', (err) => {
            console.error("Erreur spécifique lors de la lecture du CSV :");
            reject(err);
        })
        .pipe(csvParser())
        .on('data', (data) => results.push(data))
        .on('end', () => {
            console.log("Lecture CSV terminée avec succès.");
            resolve(results);
        });
});


    try {
        const [dfEvents] = await pool.execute(
            'SELECT * FROM evenement_respiratoire WHERE id_nuit = ?',
            [idNuit]
        );

        const [nbApnees, nbHypopnees, nbRera, nbrEvents] = await Promise.all([
            pool.query("CALL sp_compteur_apnees()"),
            pool.query("CALL sp_compteur_hypopnae()"),
            pool.query("CALL sp_compteur_rera()"),
            pool.query("CALL sp_compteur_all()")
        ]);

return {
            df_capteur: dfCapteur,
            df_events: dfEvents,
            nbapnees: nbApnees[0][0],
            nbhypopnees: nbHypopnees[0][0],
            nbrera: nbRera[0][0],
            nbr_events: nbrEvents[0][0]
        };
    } catch (error) {
        console.error("Erreur lors de la récupération des données:", error);
        throw error;
    }


}